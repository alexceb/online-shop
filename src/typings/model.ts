export interface ProductOption {
  color: string
  quantity: number
  power?: number[]
  storage?: string[]
}

export interface Product {
  id: number
  name: string
  brand: string
  price: string
  available: boolean
  weight: number
  options: ProductOption[]
}

export interface ProductConfiguration {
  color: string
  power?: number
  storage?: string
}

export interface CartItem extends ProductConfiguration{
  product: Product
  quantity: number
}
