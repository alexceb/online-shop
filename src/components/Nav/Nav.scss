@import '../../resources/vars.scss';

.nav {
  background-color: $shop-default-green;

  .navList {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    text-decoration: none;
    max-width: 1000px;
    margin: 0 auto;
    padding-left: 0;
  }
  
  .navListItem {
    list-style-type: none;
    padding: 10px;
    text-decoration: none;
  
    &:hover {
      background-color: $shop-dark-green;
    }
  }

  a {
    text-decoration: none;
    color: #fff;
  }
  
  .active {
    font-weight: bold;
  }
}
