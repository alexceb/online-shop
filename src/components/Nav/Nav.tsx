import * as React from 'react';
import { NavLink } from 'react-router-dom';
import * as styles from './Nav.scss';
import { CartItem } from '../../typings/model';

export interface NavigationProps {
  cart: CartItem[],
};

export const Nav: React.FC<NavigationProps> = props => {
  const { cart } = props;

  return (
    <nav className={styles.nav}>
      <ul className={styles.navList}>
        <li className={styles.navListItem}>
          <NavLink to="/" exact activeClassName={styles.active}>
            Browse products
          </NavLink>
        </li>
        <li className={styles.navListItem}>
          <NavLink to="/checkout" activeClassName={styles.active}>
            Cart{cart && cart.length ? <span>({cart.length})</span> : null}
          </NavLink>
        </li>
      </ul>
    </nav>
  )
  
};
