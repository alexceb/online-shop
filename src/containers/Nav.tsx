import { connect } from 'react-redux';
import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';

import { Nav } from '../components/Nav/Nav';
import { AppState } from '../reducers';

const mapStateToProps = (state: AppState) => ({
  cart: state.checkout.cart,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<any, any, AnyAction>) => {
  return {
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Nav);