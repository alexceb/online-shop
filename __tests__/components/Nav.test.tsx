import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CartItem, Product } from '../../src/typings/model';
import { NavigationProps, Nav } from '../../src/components/Nav/Nav';

const mockedCart: CartItem[] = [];
const mockedProps: NavigationProps = {
  cart: mockedCart,
}
const mockedProduct: Product = {
  id: 8,
  name: "Samsung 40 UHD Smart TV",
  brand: "SAMSUNG",
  price: "3200",
  available: true,
  weight: 8.2,
  options: [
    {
      color: "black",
      quantity: 19,
    }
  ],
};

const mockedCartItem: CartItem = {
  product: mockedProduct,
  quantity: 2,
  color: 'black',
  subOption: {
    optionType: null,
    optionValue: null,
  }
};

const mockedCartWithItems: CartItem[] = [mockedCartItem];

describe('Navigation component', () => {
  describe('render()', () => {
    const wrapper = shallow(<Nav {...mockedProps} />);
    test('Renders <Nav /> correctly', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('Navigation is definde and has 2 nav links', () => {
      expect(wrapper.find('.navList')).toBeDefined();
      expect(wrapper.find('.navListItem')).toHaveLength(2);
    });

    test('Should display number of cart items next to Cart NavLink', () => {
      const wrapper = shallow(<Nav cart={mockedCartWithItems} />);

      expect(wrapper.find('NavLink').find('span')).toHaveLength(1);
    });
  });
});
