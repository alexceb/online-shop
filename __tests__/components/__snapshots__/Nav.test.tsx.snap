// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component render() Renders <Nav /> correctly 1`] = `
<nav
  className="nav"
>
  <ul
    className="navList"
  >
    <li
      className="navListItem"
    >
      <NavLink
        activeClassName="active"
        exact={true}
        to="/"
      >
        Browse products
      </NavLink>
    </li>
    <li
      className="navListItem"
    >
      <NavLink
        activeClassName="active"
        to="/checkout"
      >
        Cart
      </NavLink>
    </li>
  </ul>
</nav>
`;
